samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java

```diff
@@ -704,44 +718,44 @@ public void run() {
     // start the systemConsumers for consuming input
     this.sideInputSystemConsumers.start();
 
-    // create a thread for sideInput reads
-    Thread readSideInputs = new Thread(() -> {
-        while (!shutDownSideInputRead) {
-          IncomingMessageEnvelope envelope = sideInputSystemConsumers.choose(true);
-          if (envelope != null) {
-
-            if (!envelope.isEndOfStream())
-              sideInputStorageManagers.get(envelope.getSystemStreamPartition()).process(envelope);
 
-            checkSideInputCaughtUp(envelope.getSystemStreamPartition(), envelope.getOffset(),
-                SystemStreamMetadata.OffsetType.NEWEST, envelope.isEndOfStream());
+    try {
 
-          } else {
-            LOG.trace("No incoming message was available");
+    // submit the side input read runnable
+      sideInputsReadExecutor.submit(() -> {
+          try {
+            while (!shouldShutdown) {
+              IncomingMessageEnvelope envelope = sideInputSystemConsumers.choose(true);
+
+              if (envelope != null) {
+                if (!envelope.isEndOfStream()) {
+                  sideInputStorageManagers.get(envelope.getSystemStreamPartition()).process(envelope);
+                }
+
+                checkSideInputCaughtUp(envelope.getSystemStreamPartition(), envelope.getOffset(),
+                    SystemStreamMetadata.OffsetType.NEWEST, envelope.isEndOfStream());
+              } else {
+                LOG.trace("No incoming message was available");
+              }
+            }
+          } catch (Exception e) {
+            LOG.error("Exception in reading side inputs", e);
+            sideInputException = e;
           }
-        }
-      });
+        });
 
-    readSideInputs.setDaemon(true);
-    readSideInputs.setUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
-      @Override
-      public void uncaughtException(Thread t, Throwable e) {
-        sideInputException = Optional.of(e);
-        sideInputsCaughtUp.countDown();
+      // Make the main thread wait until all sideInputs have been caughtup or an exception was thrown
+      while (sideInputException == null && !this.sideInputsCaughtUp.await(SIDE_INPUT_READ_THREAD_TIMEOUT_SECONDS,
+          TimeUnit.SECONDS)) {
+        LOG.debug("Checking for any side-input-exception occurrence");
```

Comment:
log-line seems in-complete since we are actually waiting for bootstrap to complete. 

"Waiting for SideInput bootstrap to complete"